endpoint.cpp

```diff
@@ -545,6 +545,7 @@ TcpEndpoint::TcpEndpoint()
 
 TcpEndpoint::~TcpEndpoint()
 {
+    close();
```

Comment:
You can't really call on the parent class a virtual method and expect the one on the child class to be executed.  Child class will already be gone for good when the destructor on the parent is executed.